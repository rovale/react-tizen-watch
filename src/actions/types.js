export const INITIALIZE_APP = 'INITIALIZE_APP';

export const LOAD_PAGES = 'LOAD_PAGES';
export const LOAD_DEVICES = 'LOAD_DEVICES';
export const LOAD_GROUPS = 'LOAD_GROUPS';

export const ACTIVATE_PAGE = 'ACTIVATE_PAGE';
export const SELECT_PAGE = 'SELECT_PAGE';
export const CLOSE_PAGE = 'CLOSE_PAGE';

export const ACTIVATE_DEVICE = 'ACTIVATE_DEVICE';
export const SELECT_DEVICE = 'SELECT_DEVICE';
export const CLOSE_DEVICE = 'CLOSE_DEVICE';
export const CHANGE_DEVICE_ATTRIBUTE = 'CHANGE_DEVICE_ATTRIBUTE';
export const TOGGLE_SWITCH = 'TOGGLE_SWITCH';

export const ACTIVATE_BUTTON = 'ACTIVATE_BUTTON';
export const CLICK_BUTTON = 'CLICK_BUTTON';
