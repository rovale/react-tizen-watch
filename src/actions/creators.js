import * as actionType from './types';

export const initializeApp = () => ({ type: actionType.INITIALIZE_APP });
export const handleError = (msg, url, lineNo, columnNo, error) => ({ type: actionType.HANDLE_ERROR, payload: { msg, url, lineNo, columnNo, error } });
export const loadPages = pages => ({ type: actionType.LOAD_PAGES, payload: { pages } });
export const loadDevices = devices => ({ type: actionType.LOAD_DEVICES, payload: { devices } });

export const activatePage = id => ({ type: actionType.ACTIVATE_PAGE, payload: { id } });
export const selectPage = id => ({ type: actionType.SELECT_PAGE, payload: { id } });
export const closePage = () => ({ type: actionType.CLOSE_PAGE });

export const activateDevice = id => ({ type: actionType.ACTIVATE_DEVICE, payload: { id } });
export const selectDevice = id => ({ type: actionType.SELECT_DEVICE, payload: { id } });
export const closeDevice = () => ({ type: actionType.CLOSE_DEVICE });
export const changeDeviceAttribute = event => ({ type: actionType.CHANGE_DEVICE_ATTRIBUTE, payload: { event } });

export const toggleSwitch = id => ({ type: actionType.TOGGLE_SWITCH, payload: { id } });

export const activateButton = id => ({ type: actionType.ACTIVATE_BUTTON, payload: { id } });
export const clickButton = id => ({ type: actionType.CLICK_BUTTON, payload: { id } });
