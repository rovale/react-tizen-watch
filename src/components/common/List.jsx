import React from 'react';
import PropTypes from 'prop-types';

import './List.css';

export class List extends React.Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.object).isRequired,
    activeOptionId: PropTypes.string.isRequired,
    onActivateOption: PropTypes.func.isRequired,
    onSelectOption: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);

    this.onRotaryDetent = this.onRotaryDetent.bind(this);
    this.onSelectItem = this.onSelectItem.bind(this);
  }

  componentDidMount() {
    window.addEventListener('rotarydetent', this.onRotaryDetent);
  }

  componentWillUnmount() {
    window.removeEventListener('rotarydetent', this.onRotaryDetent);
  }

  onRotaryDetent(e) {
    const { direction } = e.detail;

    const index = this.props.options.findIndex(item => item.id === this.props.activeOptionId);

    if (direction === 'CW') {
      if (index < (this.props.options.length - 1)) {
        this.props.onActivateOption(this.props.options[index + 1].id);
      }
    } else if (index > 0) {
      this.props.onActivateOption(this.props.options[index - 1].id);
    }
  }

  onSelectItem(itemId) {
    this.props.onSelectOption(itemId);
  }

  render() {
    const { options } = this.props;

    const items = options.map(item => (
      <Item
        key={item.id}
        id={item.id}
        active={item.id === this.props.activeOptionId}
        onSelect={item.onSelect || this.onSelectItem}
      >
        {item.content || <div>{item.title}</div>}
      </Item>
    ));

    return (
      <ul className="ui-listview ui-snap-listview">
        {items}
      </ul>
    );
  }
}

export class Item extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    active: PropTypes.bool.isRequired,
    onSelect: PropTypes.func.isRequired,
    children: PropTypes.element.isRequired,
  };

  static contextTypes = {
    scrollToMiddle: PropTypes.func,
  };

  constructor(props) {
    super(props);

    this.scrollToMiddle = this.scrollToMiddle.bind(this);
    this.onSelect = this.onSelect.bind(this);
    this.element = null;
  }

  componentDidMount() {
    if (this.props.active) {
      window.setTimeout(() => this.scrollToMiddle(1), 0);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.active && !this.props.active) {
      window.setTimeout(() => this.scrollToMiddle(400), 0);
    }
  }

  onSelect(e) {
    e.preventDefault();
    this.props.onSelect(this.props.id);
  }

  scrollToMiddle(duration) {
    const itemRect = this.element.getBoundingClientRect();
    this.context.scrollToMiddle(duration, itemRect.top, itemRect.height);
  }

  render() {
    const { id, active, children } = this.props;
    return (
      <li ref={(e) => { this.element = e; }} className={`ui-snap-listview-item${active ? ' ui-snap-listview-selected' : ''}`}>
        <a href={`#${id}`} onClick={this.onSelect} id={id}>{children}</a>
      </li>
    );
  }
}
