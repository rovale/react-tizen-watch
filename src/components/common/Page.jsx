import React from 'react';
import PropTypes from 'prop-types';
import scroll from 'smooth-move';

export class Page extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  };

  static childContextTypes = {
    scrollToMiddle: PropTypes.func,
  };

  constructor(props) {
    super(props);
    this.element = null;
  }

  getChildContext() {
    return { scrollToMiddle: this.scrollToMiddle };
  }

  scrollToMiddle = (duration, itemTop, itemHeight) => {
    const scroller = this.element.getElementsByClassName('ui-scroller')[0];
    const scrollerHeight = scroller.getBoundingClientRect().height;
    const newScrollTop = (itemTop + scroller.scrollTop) - ((scrollerHeight - itemHeight) / 2);

    scroll(scroller, { x: 0, y: newScrollTop, duration });
  }

  render() {
    return (
      <div ref={(e) => { this.element = e; }} className="ui-page ui-page-active ui-scroll-on">
        <div className="ui-scroller ui-snap-container" ref={e => e && e.setAttribute('tizen-circular-scrollbar', '')}>
          {this.props.children}
        </div>
      </div>
    );
  }
}

export const Header = ({ children }) => (
  <header className="ui-header">
    <h2 className="ui-title">{children}</h2>
  </header>
);

Header.propTypes = {
  children: PropTypes.string.isRequired,
};

export const Content = ({ children }) => {
  return (
    <div className="ui-content" ref={e => e && e.setAttribute('tizen-circular-scrollbar', '')}>
      {children}
    </div>);
};

Content.propTypes = {
  children: PropTypes.node.isRequired,
};
