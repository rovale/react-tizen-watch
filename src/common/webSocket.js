import io from 'socket.io-client';
import settings from './settings';
import * as action from '../actions/creators';

let socket;

export const initWebSocket = (store) => {
  const socketIoClient = window.testSocketIoClient || io;

  socket = socketIoClient(`${settings.protocol}://${settings.host}:${settings.port}/?username=${settings.userName}&password=${settings.password}`, {
    reconnection: true,
    reconnectionDelay: 1000,
    reconnectionDelayMax: 3000,
    timeout: 20000,
    forceNew: true,
  });

  socket.on('devices', (devices) => {
    store.dispatch(action.loadDevices(devices));
  });

  socket.on('pages', (pages) => {
    store.dispatch(action.loadPages(pages));
  });

  socket.on('deviceAttributeChanged', (event) => {
    if (event.deviceId !== 'syssensor' && event.deviceId !== 'temperature') {
      store.dispatch(action.changeDeviceAttribute(event));
    }
  });
};

export const toggleSwitch = (id) => {
  socket.emit('call', {
    id: 'callDeviceAction',
    action: 'callDeviceAction',
    params: {
      deviceId: id,
      actionName: 'toggle',
    },
  });
};

export const clickButton = (deviceId, buttonId) => {
  socket.emit('call', {
    id: 'callDeviceAction',
    action: 'callDeviceAction',
    params: {
      deviceId,
      buttonId,
      actionName: 'buttonPressed',
    },
  });
};
