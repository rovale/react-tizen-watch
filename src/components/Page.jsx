import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import * as action from '../actions/creators';
import { Page as UiPage, Header, Content } from './common/Page';
import { List } from './common/List';
import Switch from './deviceItem/Switch';
import Display from './deviceItem/Display';
import Presence from './deviceItem/Presence';
import Temperature from './deviceItem/Temperature';
import Thermostat from './deviceItem/Thermostat';

let Page = ({
  pageTitle, devices, activeDeviceId, onActivateDevice, onSelectDevice, onToggleSwitch,
}) => {
  const getListItemContent = (device) => {
    switch (device.template) {
      case 'device':
        return <Display device={device} />;
      case 'presence':
        return <Presence device={device} />;
      case 'switch':
        return <Switch device={device} />;
      case 'temperature':
        return <Temperature device={device} />;
      case 'thermostat':
        return <Thermostat device={device} />;
      default:
        return <div>{`${device.title} - (${device.template})`}</div>;
    }
  };

  const getOnSelect = (device) => {
    switch (device.template) {
      case 'switch':
        return () => onToggleSwitch(device.id);
      default:
        return null;
    }
  };

  const deviceOptions = devices.map(d => ({
    ...d,
    content: getListItemContent(d),
    onSelect: getOnSelect(d),
  }));

  return (
    <UiPage>
      <Header>{pageTitle}</Header>
      <Content>
        <List
          options={deviceOptions}
          activeOptionId={activeDeviceId}
          onActivateOption={onActivateDevice}
          onSelectOption={onSelectDevice}
        />
      </Content>
    </UiPage>);
};

Page.propTypes = {
  pageTitle: PropTypes.string.isRequired,
  devices: PropTypes.arrayOf(PropTypes.object).isRequired,
  activeDeviceId: PropTypes.string.isRequired,
  onActivateDevice: PropTypes.func.isRequired,
  onSelectDevice: PropTypes.func.isRequired,
  onToggleSwitch: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => {
  const selectedPage = state.pages.find(p => p.id === state.ui.selectedPageId);

  return {
    pageTitle: selectedPage.title,
    devices: selectedPage.devices.map(pd => state.devices.find(d => pd === d.id)),
    activeDeviceId: state.ui.activeDeviceId || selectedPage.devices[0],
  };
};

const mapDispatchToProps = dispatch => ({
  onActivateDevice: id => dispatch(action.activateDevice(id)),
  onSelectDevice: id => dispatch(action.selectDevice(id)),
  onToggleSwitch: id => dispatch(action.toggleSwitch(id)),
});

export default Page = connect(mapStateToProps, mapDispatchToProps)(Page);
