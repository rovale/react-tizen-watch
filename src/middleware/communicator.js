import * as actionType from '../actions/types';
import { initWebSocket, toggleSwitch, clickButton } from '../common/webSocket';

// eslint-disable-next-line arrow-parens
const communicator = store => next => action => {
  switch (action.type) {
    case actionType.INITIALIZE_APP:
      initWebSocket(store);
      break;
    case actionType.TOGGLE_SWITCH:
      toggleSwitch(action.payload.id);
      break;
    case actionType.CLICK_BUTTON: {
      const deviceId = store.getState().ui.selectedDeviceId;
      clickButton(deviceId, action.payload.id);
      break;
    }
    default:
      break;
  }

  return next(action);
};

export default communicator;
