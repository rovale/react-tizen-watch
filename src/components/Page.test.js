import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { createStore, applyMiddleware, compose } from 'redux';
import reducer from '../reducers';
import communicator from '../middleware/communicator';
import * as action from '../actions/creators';

import Page from './Page';
import { List } from './common/List';

Enzyme.configure({ adapter: new Adapter() });

const pages = [
  {
    id: 'somePage',
    name: 'Some Page',
    devices: [
      { deviceId: 'someDisplayDevice' },
      { deviceId: 'somePresenceDevice' },
      { deviceId: 'someSwitch' },
      { deviceId: 'someTemperatureDevice' },
      { deviceId: 'someThermostat' },
      { deviceId: 'someButtons' },
    ],
  },
];

/* eslint-disable object-curly-newline */
const devices = [
  { id: 'someDisplayDevice', name: 'SDD', template: 'device', attributes: [] },
  { id: 'somePresenceDevice', name: 'SPD', template: 'presence', attributes: [] },
  { id: 'someSwitch', name: 'SS', template: 'switch', attributes: [] },
  { id: 'someTemperatureDevice', name: 'STD', template: 'temperature', attributes: [] },
  { id: 'someThermostat', name: 'ST', template: 'thermostat', attributes: [] },
  { id: 'someButtons', name: 'SB', template: 'buttons', attributes: [] },
];
/* eslint-enable object-curly-newline */

describe('The Page component', () => {
  let store;
  let dispatchedActions;
  let socketIoEmitSpy;

  // eslint-disable-next-line arrow-parens, no-unused-vars
  const spyMiddleware = theStore => theNext => theAction => {
    dispatchedActions.push(theAction);
    return theNext(theAction);
  };

  beforeEach(() => {
    dispatchedActions = [];

    socketIoEmitSpy = jest.fn();
    window.testSocketIoClient = () => ({ emit: socketIoEmitSpy, on: jest.fn() });

    const enhancer = compose(applyMiddleware(spyMiddleware, communicator));
    store = createStore(reducer, enhancer);
    store.dispatch(action.initializeApp());
    store.dispatch(action.loadPages(pages));
    store.dispatch(action.loadDevices(devices));
    store.dispatch(action.selectPage('somePage'));
  });

  test('it should show a list with the devices as options', () => {
    const page = shallow((<Page store={store} />)).dive();
    const list = page.find(List);

    const listOptions = list.prop('options');
    expect(listOptions).toHaveLength(6);

    expect(list.dive()).toMatchSnapshot();
  });

  test('it should mark the first device as active option', () => {
    const page = shallow((<Page store={store} />)).dive();
    const list = page.find(List);

    expect(list.prop('activeOptionId')).toBe('someDisplayDevice');
  });

  test('it should mark the selected device as active option', () => {
    store.dispatch(action.selectDevice('someThermostat'));
    const page = shallow((<Page store={store} />)).dive();
    const list = page.find(List);

    expect(list.prop('activeOptionId')).toBe('someThermostat');
  });

  test('it should activate the device when the option is activated in the list', () => {
    const page = shallow((<Page store={store} />)).dive();
    const list = page.find(List);

    list.prop('onActivateOption')('someThermostat');

    expect(store.getState().ui.activeDeviceId).toBe('someThermostat');
    expect(store.getState().ui.selectedDeviceId).toBeNull();
  });

  test('it should select the device when the option is selected in the list', () => {
    const page = shallow((<Page store={store} />)).dive();
    const list = page.find(List);
    const listContent = list.dive();
    const listItem = listContent.find('Item[id="someThermostat"]');

    // TODO: make onSelect work without specifying the id.
    listItem.prop('onSelect')('someThermostat');

    expect(store.getState().ui.activeDeviceId).toBe('someThermostat');
    expect(store.getState().ui.selectedDeviceId).toBe('someThermostat');
  });

  test('it should toggle the switch when the option is a switch', () => {
    const page = shallow((<Page store={store} />)).dive();
    const list = page.find(List);
    const listContent = list.dive();
    const listItemWithSwitch = listContent.find('Item[id="someSwitch"]');

    listItemWithSwitch.prop('onSelect')();

    expect(dispatchedActions[dispatchedActions.length - 1]).toMatchObject(action.toggleSwitch('someSwitch'));
    expect(socketIoEmitSpy.mock.calls).toMatchSnapshot();
  });
});
