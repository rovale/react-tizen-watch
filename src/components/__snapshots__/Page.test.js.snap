// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Page component it should show a list with the devices as options 1`] = `
<ul
  className="ui-listview ui-snap-listview"
>
  <Item
    active={true}
    id="someDisplayDevice"
    key="someDisplayDevice"
    onSelect={[Function]}
  >
    <Display
      device={
        Object {
          "attributes": Array [],
          "config": undefined,
          "id": "someDisplayDevice",
          "template": "device",
          "title": "SDD",
        }
      }
    />
  </Item>
  <Item
    active={false}
    id="somePresenceDevice"
    key="somePresenceDevice"
    onSelect={[Function]}
  >
    <Presence
      device={
        Object {
          "attributes": Array [],
          "config": undefined,
          "id": "somePresenceDevice",
          "template": "presence",
          "title": "SPD",
        }
      }
    />
  </Item>
  <Item
    active={false}
    id="someSwitch"
    key="someSwitch"
    onSelect={[Function]}
  >
    <Switch
      device={
        Object {
          "attributes": Array [],
          "config": undefined,
          "id": "someSwitch",
          "template": "switch",
          "title": "SS",
        }
      }
    />
  </Item>
  <Item
    active={false}
    id="someTemperatureDevice"
    key="someTemperatureDevice"
    onSelect={[Function]}
  >
    <Temperature
      device={
        Object {
          "attributes": Array [],
          "config": undefined,
          "id": "someTemperatureDevice",
          "template": "temperature",
          "title": "STD",
        }
      }
    />
  </Item>
  <Item
    active={false}
    id="someThermostat"
    key="someThermostat"
    onSelect={[Function]}
  >
    <Thermostat
      device={
        Object {
          "attributes": Array [],
          "config": undefined,
          "id": "someThermostat",
          "template": "thermostat",
          "title": "ST",
        }
      }
    />
  </Item>
  <Item
    active={false}
    id="someButtons"
    key="someButtons"
    onSelect={[Function]}
  >
    <div>
      SB - (buttons)
    </div>
  </Item>
</ul>
`;

exports[`The Page component it should toggle the switch when the option is a switch 1`] = `
Array [
  Array [
    "call",
    Object {
      "action": "callDeviceAction",
      "id": "callDeviceAction",
      "params": Object {
        "actionName": "toggle",
        "deviceId": "someSwitch",
      },
    },
  ],
]
`;
